_posts/2018-01-14-smaller-docker-images.md

```diff
@@ -0,0 +1,405 @@
+
+You don't need any of those when you run your container. The only dependency you need is Node.js.
+
+You could remove everything but Node.js. But how?
+
+Fortunately Google had the same idea and came up with [GoogleCloudPlatform/distroless](https://github.com/GoogleCloudPlatform/distroless).
+
+As the description for the repository points out:
+
+> "Distroless" images contain only your application and its runtime dependencies. They do not contain package managers, shells any other programs you would expect to find in a standard Linux distribution.
+
+This is exaclty what you need!
+
+You can tweak the `Dockerfile` to leverage the new base image like this:
+
+```dockerfile
+FROM node:8 as build
+
+WORKDIR /app
+COPY package.json .
+RUN yarn install
+COPY index.js .
+
+FROM gcr.io/distroless/nodejs
+
+COPY --from=build /app /
+EXPOSE 3000
+CMD ["index.js"]
+```
+
+And you can compile the image as usual with:
+
+```bash
+$ docker build -t node-distroless .
+```
+
+The application should run as normal. To verify that is still the case, you could run the container like this:
+
+```bash
+$ docker run -p 3000:3000 node-distroless
+```
+
+And visit the page at [http://localhost:3000](http://localhost:3000).
+
+Is the image without all the extra binaries smaller?
+
+```bash
+$ docker images | grep node-distroless
+node-distroless                 latest              7b4db3b7f1e5        3 minutes ago       76.7MB
+```
+
+**That's only 76.7MB!**
+
+More than 600MB down from your previous image!
+
+This is extremely good news, but there's something you should pay attention to when it comes to distroless.
+
+When you container is running and you wish to inspect it, you can attach to a running container with:
+
+```bash
+$ docker exec -ti <inser_docker_id> bash
+```
+
+Attaching to a running container and running `bash` feels like establing an SSH session.
+
+But since distroless is a stripped down version of the original operating system, there are no extra binaries. There's no shell in the container!
+
+How can you attach to a running container if there's no shell?
+
+The good and the bad news is that you can't.
+
+It's bad news because you can only execute the binaries in the container. The only binary you could use is Node.js:
+
+```bash
+$ docker exec -ti <inser_docker_id> node
+```
+
+It's good news because an attacker exploiting your application and gaining access to the container won't be able to do as much damage as if were to access a shell. In other words, less binaries means smaller sizes and increased security. But at the cost of more painful debugging.
+
+But what if you cared about debugging and smaller sizes?
+
+## 3. Smaller base images with Alpine
+
+You could replace the distroless base image with an Alpine based image.
+
+Alpine Linux is:
+
+> a security-oriented, lightweight Linux distribution based on musl libc and busybox
```

Comment:
based on [musl libc](https://www.musl-libc.org/) and [busybox](https://www.busybox.net/)